src/tiled/moveselectiontool.cpp

```diff
@@ -0,0 +1,184 @@
+/*
+ * moveselectiontool.cpp
+ * Copyright 2017, Ketan Gupta <ketan19972010@gmail.com>
+ *
+ * This file is part of Tiled.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "moveselectiontool.h"
+
+#include "brushitem.h"
+#include "changeselectedarea.h"
+#include "erasetiles.h"
+#include "mapdocument.h"
+#include "painttilelayer.h"
+#include "tilesetmanager.h"
+
+#include <QApplication>
+#include <QUndoStack>
+
+using namespace Tiled;
+using namespace Tiled::Internal;
+
+MoveSelectionTool::MoveSelectionTool(QObject *parent)
+    : AbstractTileTool(tr("Move Selection"),
+                       QIcon(QLatin1String(
+                               ":images/22x22/move-selection.png")),
+                       QKeySequence(tr("G")),
+                       parent)
+    , mDragging(false)
+    , mMouseDown(false)
+{
+}
+
+MoveSelectionTool::~MoveSelectionTool()
+{
+}
+
+void MoveSelectionTool::activate(MapScene *scene)
+{
+    AbstractTileTool::activate(scene);
+    cut();
+}
+
+void MoveSelectionTool::deactivate(MapScene *scene)
+{
+    AbstractTileTool::deactivate(scene);
+    paste();
+}
+
+void MoveSelectionTool::tilePositionChanged(const QPoint &pos)
+{
+    if (mDragging) {
+        QPoint offset = pos - mLastUpdate;
+        QRegion selectedArea = mapDocument()->selectedArea();
+        selectedArea.translate(offset);
+
+        mapDocument()->undoStack()->push(new ChangeSelectedArea(mapDocument(), selectedArea));
+
+        if (mPreviewLayer) {
+            mPreviewLayer->setX(mPreviewLayer->x()+offset.x());
+            mPreviewLayer->setY(mPreviewLayer->y()+offset.y());
+            brushItem()->setTileLayer(mPreviewLayer);
+            brushItem()->setTileRegion(mapDocument()->selectedArea());
```

Comment:
Please use the version of `setTileLayer` that takes the region as second parameter. Also, since we shouldn't rely on the selected area, you could instead translate the current tile region by the offset as well: `brushItem()->tileRegion().translated(offset)`.